src/compiler/checker.ts

```diff
@@ -24121,6 +24111,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
+        const sourceValueDeclaration = sourceType.symbol?.valueDeclaration;
+        if (sourceValueDeclaration) {
+            const intraExpressionInferenceSites = getInferenceContext(sourceValueDeclaration)?.intraExpressionInferenceSites?.filter(site => isNodeDescendantOf(site.node, sourceValueDeclaration));
+            if (intraExpressionInferenceSites?.length) {
+                const templateType = (getApparentTypeOfContextualType(sourceValueDeclaration.parent.parent as Expression, ContextFlags.NoConstraints) as MappedType).templateType;
+                if (templateType) {
+                    Debug.assert(isExpressionNode(sourceValueDeclaration));
+                    pushContextualType(sourceValueDeclaration as any as Expression, templateType, /*isCache*/ false);
+                    inferFromIntraExpressionSites([inference], intraExpressionInferenceSites);
+                    popContextualType();

```

Comment:
At first this felt quite hacky but now I'm leaning towards saying that it's the easiest and cleanest solution to this problem. The problem that this is solving (together with the `getApparentTypeOfContextualType(sourceValueDeclaration.parent.parent)` call a few lines back is that it **avoids** instantiating the property name. Essentially it allows me use the bare `templateType` without substituting indexed mapped types so `inferFromIntraExpressionSites` can see a target type like `(x: string) => T[K]` instead of `(x: string) => T["a"]` and thus it can infer things for `T[K]`.

I wasn't sure what value I should pass as `isCache` here though. I'm pretty sure that @ahejlsberg would know this and could advise here